patches/api/0387-Paper-Plugins.patch

```diff
@@ -0,0 +1,963 @@
++            if (checkGlobal || result.getClassLoader() == this) {
++                return result;
++            }
++        } catch (ClassNotFoundException ex) {
++        }
++
++        if (checkGlobal) {
++            // This ignores the libraries of other plugins, unless they are transitive dependencies.
++            Class<?> result = this.loader.getClassByName(name, resolve, this.pluginConfiguration, this);  // Paper - prioritize self
++
++            if (result != null) {
++                // If the class was loaded from a library instead of a PluginClassLoader, we can assume that its associated plugin is a transitive dependency and can therefore skip this check.
++                if (result.getClassLoader() instanceof PaperPluginClassLoader existingClassLoader) {
++                    PluginConfiguration classLoaderConfig = existingClassLoader.pluginConfiguration;
++
++                    String identifier = classLoaderConfig.getIdentifier();
++                    if (classLoaderConfig != this.pluginConfiguration && !this.seenIllegalAccess.contains(identifier) && !this.pluginConfiguration.isTransitiveDependency(classLoaderConfig)) {
++                        this.seenIllegalAccess.add(identifier);
++                        if (this.pluginProvider != null) {
++                            this.logger.warn("Loaded class {} from {} which is not a depend or softdepend of this plugin.", name, classLoaderConfig.getDisplayName());
++                        }
++                    }
++                }
++
++                return result;
++            }
++        }
++
++        throw new ClassNotFoundException(name);
++    }
++
++    @Override
++    protected Class<?> findClass(String name) throws ClassNotFoundException {
++        if (name.startsWith("org.bukkit.") || name.startsWith("net.minecraft.")) {
++            throw new ClassNotFoundException(name);
++        }
++        Class<?> result = this.loadedClasses.get(name);
++
++        if (result == null) {
++            String path = name.replace('.', '/').concat(".class");
++            JarFile jar = this.pluginProvider.file();
++            JarEntry entry = jar.getJarEntry(path);
++
++            if (entry != null) {
++                byte[] classBytes;
++
++                try (InputStream is = jar.getInputStream(entry)) {
++                    classBytes = ByteStreams.toByteArray(is);
++                } catch (IOException ex) {
++                    throw new ClassNotFoundException(name, ex);
++                }
++
++                //classBytes = loader.server.getUnsafe().processClass(description, path, classBytes); TODO: Awful byte manipulation
++
++                int dot = name.lastIndexOf('.');
++                if (dot != -1) {
++                    String pkgName = name.substring(0, dot);
++                    // TODO: Do we want to use the deprecated package method here?
```

Comment:
Well it depends if you have multiple class loader or not but generally keep the undeprecated one